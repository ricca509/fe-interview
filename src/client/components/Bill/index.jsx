import React, { Component } from 'react';
import PropTypes from 'prop-types';
import Transactions from '../Transactions';

class Bill extends Component {
  constructor() {
    super();

    this.state = {
      expanded: false
    };

    this.onNameClicked = this.onNameClicked.bind(this);
  }

  onNameClicked() {
    this.setState(() => ({ expanded: !this.state.expanded }));
  }

  render() {
    const { name, id, transactions, isBill, onActionClicked } = this.props;
    const { expanded } = this.state;
    const buttonCopy = isBill ? 'Remove bill' : 'Add as bill';
    const onBillSelected = evt => onActionClicked(id);

    return (
      <li className="bill">
        <span onClick={this.onNameClicked}>{name}</span>
        <button className="bill-action" onClick={onBillSelected}>
          {buttonCopy}
        </button>
        { expanded && <Transactions transactions={transactions} />}
      </li>
    );
  }
};

Bill.propTypes = {
  name: PropTypes.string,
  id: PropTypes.string,
  transactions: PropTypes.array,
  isBill: PropTypes.bool,
  onSelected: PropTypes.func
};

export default Bill;