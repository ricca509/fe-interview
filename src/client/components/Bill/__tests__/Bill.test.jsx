import React from 'react';
import Bill from '../';
import { shallow } from 'enzyme';

describe('Bill component', ()=> {
  it('should render a list element with the bill name and a button with the right action', () => {
    const props = {
      name: 'Vodafone',
      isBill: false,
      transactions: [
        {
          amount: 12.34,
          date: '2017-01-13'
        },
        {
          amount: 14.34,
          date: '2017-02-13'
        }
      ],
      id: '5a5caa1efe33900100fd8ed5'
    };

    const wrapper = shallow(<Bill {...props} />);

    expect(wrapper).toMatchSnapshot();
  });

  it('should show `Transactions` when clicked', () => {
    const props = {
      name: 'Vodafone',
      isBill: false,
      transactions: [
        {
          amount: 12.34,
          date: '2017-01-13'
        },
        {
          amount: 14.34,
          date: '2017-02-13'
        }
      ],
      id: '5a5caa1efe33900100fd8ed5'
    };

    const wrapper = shallow(<Bill {...props} />);
    wrapper.find('span').simulate('click');

    expect(wrapper.find('Transactions')).toMatchSnapshot();
  });

  it('should call the callback with the transaction id when the button is clicked', () => {
    const props = {
      name: 'Vodafone',
      isBill: false,
      transactions: [
        {
          amount: 12.34,
          date: '2017-01-13'
        },
        {
          amount: 14.34,
          date: '2017-02-13'
        }
      ],
      id: '5a5caa1efe33900100fd8ed5',
      onActionClicked: jest.fn()
    };

    const wrapper = shallow(<Bill {...props} />);
    wrapper.find('button').simulate('click');

    expect(props.onActionClicked).toHaveBeenCalledWith('5a5caa1efe33900100fd8ed5');
  });
});