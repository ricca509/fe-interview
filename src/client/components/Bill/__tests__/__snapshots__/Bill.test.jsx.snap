// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Bill component should render a list element with the bill name and a button with the right action 1`] = `
<li
  className="bill"
>
  <span
    onClick={[Function]}
  >
    Vodafone
  </span>
  <button
    className="bill-action"
    onClick={[Function]}
  >
    Add as bill
  </button>
</li>
`;

exports[`Bill component should show \`Transactions\` when clicked 1`] = `
<Transactions
  transactions={
    Array [
      Object {
        "amount": 12.34,
        "date": "2017-01-13",
      },
      Object {
        "amount": 14.34,
        "date": "2017-02-13",
      },
    ]
  }
/>
`;
